src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/CallStats.cs

```diff
@@ -0,0 +1,205 @@
+                SunnyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            catch (Exception e)
+            {
+                RainyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+        }
+
+        public T CallFuncAndRecordStats<T>(Func<T> func)
+        {
+            T result = default(T);
+            Stopwatch sw = new Stopwatch();
+            try
+            {
+                sw.Restart();
+                result = func();
+                SunnyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            catch (Exception e)
+            {
+                RainyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            return result;
+        }
+
+        public T CallFuncAndRecordStats<T, P>(Func<P, T> func, P param)
+        {
+            T result = default(T);
+            Stopwatch sw = new Stopwatch();
+            try
+            {
+                sw.Restart();
+                result = func(param);
+                SunnyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            catch (Exception e)
+            {
+                RainyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            return result;
+        }
+
+        public async Task CallAsyncFuncAndRecordStatsAsync(Func<Task> func)
+        {
+            Stopwatch sw = new Stopwatch();
+            try
+            {
+                sw.Restart();
+                await func();
+                SunnyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            catch (Exception e)
+            {
+                RainyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+        }
+
+        public async Task CallAsyncFuncAndRecordStatsAsync<P>(Func<P, Task> func, P param)
+        {
+            Stopwatch sw = new Stopwatch();
+            try
+            {
+                sw.Restart();
+                await func(param);
+                SunnyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+            catch (Exception e)
+            {
+                RainyDay.AddCallTiming(sw.ElapsedTicks);
+            }
+        }
+    }
+
+    public class TimingPercentileStats
+    {
+        public DateTime Time;       // Time when the stats were calculated
+        public int[] Centile;       // negative % means take bottom part, positive % means take top part
+        public long[] AvgTiming;    // within this time
+    }
+
+    public class CallTimingStatsCollector
+    {
+        private int _samples;
+        private int _numCalls;
+        private long[] _timings;
+        private Object _lock = new Object();
+        private Func<long[], TimingPercentileStats> _statsCalculator;
+        public ConcurrentQueue<TimingPercentileStats> PercentileStats { get; set; }
+
+        public CallTimingStatsCollector(int samples, Func<long[], TimingPercentileStats> statsCalculator)
+        {
+            _samples = samples;
+            _timings = new long[_samples * 2];
+            _statsCalculator = statsCalculator;
```

Comment:
CallTimingStatsCollector was "invented" as a short-term substitution of a much more thorough and hopefully well thought out solution for collecting and analyzing the timings that David/Jorge are working on in their stress infrastructure. I wanted this temporary solution to be:
- small / compact and not over engineered (and also not a lot of time investment)
- have no dependencies on anything else (ETW, ILogger, etc)
- fast (preferably lock-free) so that we won't have additional contentions / serialization of execution in our stress tests
- low memory & CPU footprint (e.g. not many allocations so we can focus on the product instead)
- to give a fairly good representation of the collected data

Now, to your question: with the number of samples typically over 10K we can simply overcommit the amount of memory we allocated for those samples and *assume* that we won't collect more than _samples while we process the already collected _samples. When this assumption is broken (_samples is small or the rate of adding new time samples is very high) we will loose some data and the stats won't be precise. The No1 reason for this approach is its code simplicity, although it is not that difficult to design a loss-less solution similar to the one in SignalR with a circular array of buckets to store messages. 

I have some more concerns over this code too:
- the allocation of long[] copy may land to LOH - we might consider making it a the stack allocation (if the size isn't too large comparing to the remaining stack size).
- the Func<long[], TimingPercentileStats> statsCalculator should be generic towards the result it produces - e.g. different stats calculators should be able to produce totally different representation of the data it collects in real-time. Re-working the stats calculation would also eliminate the non test-specific magic numbers like (-50 and 1) that you mentioned below.
- Since we're collecting the data real-time we may also benefit from catching (and breaking into debugger ) unusually long-running operations (which I actually do see in stress based on the stats that I collected). 

In this initial check-in I tried to keep the code related to the stats to the minimum as there are many possible directions to improve it.